openhands/mcp/utils.py

```diff
@@ -155,8 +161,30 @@ async def call_tool_mcp(mcp_clients: list[MCPClient], action: MCPAction) -> Obse
     )
 
 
+def add_search_engine(app_config: AppConfig) -> MCPStdioServerConfig | None:
+    """Add search engine to the MCP config"""
+    from openhands.server.shared import server_config
+
+    if server_config.app_mode == AppMode.OSS:
+        if (
+            app_config.search_api_key
+            and app_config.search_api_key.get_secret_value().startswith('tvly-')
+        ):
+            logger.info('Adding search engine to MCP config')
+            return MCPStdioServerConfig(
+                name='tavily',
+                command='npx',
+                args=['-y', 'tavily-mcp@0.1.4'],
+                env={'TAVILY_API_KEY': app_config.search_api_key.get_secret_value()},
+            )
+        else:
+            logger.warning('No search engine API key found, skipping search engine')
+    # Do not add search engine to MCP config in SaaS mode since it will be added by the OpenHands server
+    return None
```

Comment:
Hmm seems like an unideal pattern to use server components here (it would be better to keep the server self contained)

Alternatively, I think we could relocate this method to [`OpenHandsMCPConfig` ](https://github.com/All-Hands-AI/OpenHands/blob/main/openhands/core/config/mcp_config.py#L144), and always return the `McpStudioServerConfig`

Then override the method in saas to always  return `None`; we should have the same behavior and it would organize all default mcp server configs of openhands in one place

Reply:
> I think we could relocate this method to [OpenHandsMCPConfig ](https://github.com/All-Hands-AI/OpenHands/blob/main/openhands/core/config/mcp_config.py#L144), and always return the McpStudioServerConfig

I think the tricky part is that the user might not have set `app_config.search_api_key` the moment we initialize `OpenHandsMCPConfig` :(

Reply:
Maybe we append this server to the `mcp` field in `AppConfig` when initializing the runtime then? The runtime should have the latest set config?